File: bindings/nodejs/examples/how_tos/sign_ed25519/sign_ed25519.ts
// Copyright 2023 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {
    CoinType,
    HD_WALLET_TYPE,
    initLogger,
    SecretManager,
    utf8ToHex,
    Utils,
} from '@iota/sdk';
require('dotenv').config({ path: '.env' });

// In this example we will sign with Ed25519.
// Run with command:
// node ./dist/how_tos/sign_ed25519/sign_ed25519.js

const FOUNDRY_METADATA = {
    standard: 'IRC30',
    name: 'NativeToken',
    description: 'A native token',
    symbol: 'NT',
    decimals: 6,
    logoUrl: 'https://my.website/nativeToken.png',
};

async function run() {
    initLogger();

    try {
        if (!process.env.STRONGHOLD_PASSWORD) {
            throw new Error(
                '.env stronghold password is undefined, see .env.example',
            );
        }
        if (!process.env.NON_SECURE_USE_OF_DEVELOPMENT_MNEMONIC_1) {
            throw new Error('.env mnemonic is undefined, see .env.example');
        }
        const secretManager = new SecretManager({
            stronghold: {
                password: process.env.STRONGHOLD_PASSWORD,
                snapshotPath: 'sign_ed25519.stronghold',
            },
        });

        // A mnemonic can be generated with `Utils.generateMnemonic()`.
        // Store the mnemonic in the Stronghold snapshot, this needs to be done only the first time.
        // The mnemonic can't be retrieved from the Stronghold file, so make a backup in a secure place!
        await secretManager.storeMnemonic(
            process.env.NON_SECURE_USE_OF_DEVELOPMENT_MNEMONIC_1,
        );

        const message = utf8ToHex(JSON.stringify(FOUNDRY_METADATA));
        const ed25519Signature = await secretManager.signEd25519(message, [
            HD_WALLET_TYPE,
            CoinType.Shimmer,
            0,
            0,
            0,
        ]);
        console.log(
            `Public key: ${ed25519Signature.publicKey}\nSignature: ${ed25519Signature.signature}`,
        );

        const bech32Address = Utils.hexPublicKeyToBech32Address(
            ed25519Signature.publicKey,
            'rms',
        );
        console.log('Address: ' + bech32Address);
    } catch (error) {
        console.error('Error: ', error);
    }
}

run().then(() => process.exit());

File: sdk/examples/how_tos/sign_and_verify_ed25519/sign_ed25519.rs
// Copyright 2024 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

//! In this example we will sign with Ed25519.
//!
//! Rename `.env.example` to `.env` first, then run the command:
//! ```sh
//! cargo run --release --all-features --example sign_ed25519
//! ```

use crypto::keys::bip44::Bip44;
use iota_sdk::{
    client::{
        constants::SHIMMER_COIN_TYPE,
        hex_public_key_to_bech32_address,
        secret::{stronghold::StrongholdSecretManager, SecretManage, SecretManager},
    },
    crypto::keys::bip39::Mnemonic,
};

const FOUNDRY_METADATA: &str = r#"{"standard":"IRC30","name":"NativeToken","description":"A native token","symbol":"NT","decimals":6,"logoUrl":"https://my.website/nativeToken.png"}"#;
const ACCOUNT_INDEX: u32 = 0;
const INTERNAL_ADDRESS: bool = false;
const ADDRESS_INDEX: u32 = 0;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // This example uses secrets in environment variables for simplicity which should not be done in production.
    dotenvy::dotenv().ok();

    for var in ["STRONGHOLD_PASSWORD", "MNEMONIC"] {
        std::env::var(var).expect(&format!(".env variable '{var}' is undefined, see .env.example"));
    }

    // Setup Stronghold secret_manager
    let stronghold = StrongholdSecretManager::builder()
        .password(std::env::var("STRONGHOLD_PASSWORD").unwrap())
        .build("sign_ed25519.stronghold")?;

    stronghold
        .store_mnemonic(Mnemonic::from(std::env::var("MNEMONIC").unwrap()))
        .await?;

    let bip44_chain = Bip44::new(SHIMMER_COIN_TYPE)
        .with_account(ACCOUNT_INDEX)
        .with_change(INTERNAL_ADDRESS as _)
        .with_address_index(ADDRESS_INDEX);

    let message = FOUNDRY_METADATA.as_bytes();
    let signature = SecretManager::Stronghold(stronghold)
        .sign_ed25519(message, bip44_chain)
        .await?;
    println!(
        "Public key: {}\nSignature: {}",
        prefix_hex::encode(signature.public_key_bytes().as_ref()),
        prefix_hex::encode(signature.signature().to_bytes()),
    );

    // Hash the public key to get the address
    let bech32_address =
        hex_public_key_to_bech32_address(&prefix_hex::encode(signature.public_key_bytes().as_ref()), "rms")?;
    println!("Address: {bech32_address}");

    Ok(())
}
